### ISSUE TITLE
S3 state provider will remove stack data on connection lost while `pulumi up`

### ISSUE THREAD

**Issue by ypresto** [pulumi/pulumi#4679]
Not so long ago, my `state.json` is deleted on connection lost while `pulumi up`. Fortunately I have `state.json.bak` and turned on S3 versioning, I could restore it.

Because there are `Delete Marker`s in version history of `state.json`, I concluded the state file is explicitly deleted before update.

We can just overwrite it without deleting it..?

**Comment by theneva** [comment]
Hi! :wave: This exact issue happened to me just now using Google Cloud Storage as a backend. cc @leezen (if you want to update this issue's entry in #4605)

I somehow hit the rate limit for the `dev` stack file:

`error: pre-step event returned an error: failed to save snapshot: An IO error occurred during the current operation: blob (key ".pulumi/stacks/dev.json") (code=Unknown): googleapi: Error 429: The rate of change requests to the object <redacted bucket name>/.pulumi/stacks/dev.json exceeds the rate limit. Please reduce the rate of create, update, and delete requests., rateLimitExceeded`

In other words, Pulumi managed to move `dev.json` to `dev.json.bak`, but failed to write the new snapshot.

Fortunately restoring the stack was just a matter of copying `dev.json.bak` to `dev.json` and re-running `pulumi up`, but I was pretty stressed out for a couple of minutes :smile:

Do you think the fact that I hit the rate limit deserves its own issue?

**Comment by theneva** [comment]
Ah! I was on 1.x when it happened, but upgraded to 2.x literally seconds later as part of debugging it... I guess you can disregard my case :smile:

**Comment by JakeGinnivan** [comment]
I think this may be contributing to stack references missing when one project is deploying then another tries to access the stack references.

**Comment by reddoggad** [comment]
It happened to me too! (after ctrl+c)

pulumi version:
```
v3.2.1
```

Stack trace:
```
Duration: 33s

panic: send on closed channel

goroutine 17693 [running]:
github.com/pulumi/pulumi/pkg/v3/engine.diagEvent(0xc001caaa50, 0xc00185cd80, 0xc0021d0d80, 0x1c, 0xc001c3b8c0, 0x51, 0x1d2a672, 0x5, 0x0)
        /Users/runner/work/pulumi/pulumi/pkg/engine/events.go:578 +0x20c
github.com/pulumi/pulumi/pkg/v3/engine.(*eventEmitter).diagErrorEvent(...)
        /Users/runner/work/pulumi/pulumi/pkg/engine/events.go:602
github.com/pulumi/pulumi/pkg/v3/engine.(*eventSink).Errorf(0xc001caaa50, 0xc00185cd80, 0x0, 0x0, 0x0)
        /Users/runner/work/pulumi/pulumi/pkg/engine/eventsink.go:88 +0x127
github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*stepExecutor).executeChain(0xc000ca2500, 0x38, 0xc002a06dc0, 0x1, 0x1)
        /Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step_executor.go:231 +0x38a
github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*stepExecutor).worker.func1(0xc000ca2500, 0x38, 0xc002a06dc0, 0x1, 0x1, 0xc001c3be00)
        /Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step_executor.go:377 +0xc5
created by github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*stepExecutor).worker
        /Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step_executor.go:374 +0x2c8
panic: send on closed channel

goroutine 17941 [running]:
github.com/pulumi/pulumi/pkg/v3/engine.diagEvent(0xc001caaa50, 0xc00196f650, 0xc001a4f5c0, 0x1c, 0xc001a56240, 0x51, 0x1d2a672, 0x5, 0x0)
        /Users/runner/work/pulumi/pulumi/pkg/engine/events.go:578 +0x20c
github.com/pulumi/pulumi/pkg/v3/engine.(*eventEmitter).diagErrorEvent(...)
        /Users/runner/work/pulumi/pulumi/pkg/engine/events.go:602
github.com/pulumi/pulumi/pkg/v3/engine.(*eventSink).Errorf(0xc001caaa50, 0xc00196f650, 0x0, 0x0, 0x0)
        /Users/runner/work/pulumi/pulumi/pkg/engine/eventsink.go:88 +0x127
github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*stepExecutor).executeChain(0xc000ca2500, 0x3c, 0xc00207d5b0, 0x1, 0x1)
        /Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step_executor.go:231 +0x38a
github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*stepExecutor).worker.func1(0xc000ca2500, 0x3c, 0xc00207d5b0, 0x1, 0x1, 0xc001dc0420)
        /Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step_executor.go:377 +0xc5
created by github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*stepExecutor).worker
        /Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step_executor.go:374 +0x2c8
```
i hope it will help u.

**Comment by NickLarsenNZ** [comment]
Just happened to me on `v3.15.0` :/

**Comment by introom** [comment]
> We just had that issue in version 3 on Azure Storage. Sadly I don't have the log anymore concerning that issue. Perhaps I can try to reproduce it if it would help. It happened while the terminal session died during `pulumi up`

Also having this on azure storage. Pulumi version 3.20

**Comment by introom** [comment]
Reproducing this issue is quite straightforward, for example,
Try install the cert-manager helm chart, and hitting several Ctrl-c in the middle to force quit.

**Comment by dkrieger** [comment]
it is extremely concerning this issue has been open for 3.5 years. I found this issue while trying to figure out why pulumi history version appears to be 0 for everything returned by `pulumi stack history` (not sure if it's caused by this issue, but I have a hunch it is). using digitalocean s3 compatible backend, but based on other answers, it appears this affects all s3 compatible backends. I'm on the latest pulumi version, 3.105.0

deleting the state file instead of overwriting it in place is insane behavior at face value -- can we get some movement or at least some meaningful commentary on this issue @mikhailshilkov @EvanBoyle ?

**Comment by iwahbe** [comment]
This was fixed by https://github.com/pulumi/pulumi/pull/9122.

@dkrieger if you hit a problem after https://github.com/pulumi/pulumi/pull/9122 merged & released, please open a new issue.
